### PR TITLE
Master clean reinvoice jem

### PR DESCRIPTION
**Pull request by jem-odoo**
https://www.odoo.com/web?debug#id=&action=333&active_id=965&model=project.task&view_type=form&menu_id=4720

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by jem-odoo** [comment]
@qdp-odoo @csnauwaert I might need your input here on 2 things:
1/ are you ok to create analytic line in batch when posting a move.line ? (first commit)
2/ Will it cause problem (2nd commit) for the account.move.line to create the sale order line and the analytic lines ? Instead of doing it the AAL creation. Is that shocking ? :)
Thanks ;)

File: addons/sale/models/account_move.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+            if 'so_line' not in values and (move_line.credit or 0.0) <= (move_line.debit or 0.0) and move_line.product_id.expense_policy not in [False, 'no']:
+                move_to_reinvoice |= move_line
+
+        # insert the sale line in the create values of the analytic entries
+        if move_to_reinvoice:
+            map_sale_line_per_move = move_to_reinvoice._sale_create_reinvoice_sale_line()
+
+            for values in values_list:
+                move_line_id = values.get('move_id')
+                sale_line = map_sale_line_per_move.get(move_line_id)
```
Comment (dbeguin) [inline addons/sale/models/account_move.py]: move_line_id : useless variable
 sale_line = map_sale_line_per_move.get(values.get('move_id'))

File: addons/sale/models/account_move.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+            price = move_line._sale_get_invoice_price(sale_order)
+
+            # find the existing sale.line or keep its creation values to process this in batch
+            sale_line = None
+            if move_line.product_id.expense_policy == 'sales_price' and move_line.product_id.invoice_policy == 'delivery':  # for those case only, we can try to reuse one
+                map_entry_key = (sale_order.id, move_line.product_id.id, price)  # cache entry to limit the call to search
+                sale_line = existing_sale_line_cache.get(map_entry_key)
+                if sale_line:  # already search, so reuse it. sale_line can be sale.order.line record or index of a "to create values" in `sale_line_values_to_create`
+                    map_move_sale_line[move_line.id] = sale_line
+                    existing_sale_line_cache[map_entry_key] = sale_line
```
Comment (dbeguin) [inline addons/sale/models/account_move.py]: As you can easily understand with the comment, you can directly assign both values
`map_move_sale_line[move_line.id] = existing_sale_line_cache[map_entry_key] = sale_line`

File: addons/sale/models/account_move.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                    if sale_line:  # found existing one, so keep the browse record
+                        map_move_sale_line[move_line.id] = sale_line
+                        existing_sale_line_cache[map_entry_key] = sale_line
+                    else:  # should be create, so use the index of creation values instead of browse record
+                        # save value to create it
+                        sale_line_values_to_create.append(move_line._sale_prepare_sale_line_values(sale_order, price))
+                        # store it in the cache of existing ones
+                        existing_sale_line_cache[map_entry_key] = len(sale_line_values_to_create) - 1  # save the index of the value to create sale line
+                        # store it in the map_move_sale_line map
+                        map_move_sale_line[move_line.id] = len(sale_line_values_to_create) - 1  # save the index of the value to create sale line
```
Comment (dbeguin) [inline addons/sale/models/account_move.py]: same principle

File: addons/sale/models/account_move.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+                else:  # search for existing sale line
+                    sale_line = self.env['sale.order.line'].search([
+                        ('order_id', '=', sale_order.id),
+                        ('price_unit', '=', price),
+                        ('product_id', '=', move_line.product_id.id),
+                        ('is_expense', '=', True),
+                    ], limit=1)
+                    if sale_line:  # found existing one, so keep the browse record
+                        map_move_sale_line[move_line.id] = sale_line
+                        existing_sale_line_cache[map_entry_key] = sale_line
```
Comment (dbeguin) [inline addons/sale/models/account_move.py]: same remark

File: addons/sale/models/account_move.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+            :return a dict where key is the move line id, and value is sale.order record (or None).
+        """
+        analytic_accounts = self.mapped('analytic_account_id')
+
+        # link the analytic account with its open SO by creating a map: {AA.id: sale.order}, if we find some analytic accounts
+        mapping = {}
+        if analytic_accounts:  # first, search for the open sales order
+            sale_orders = self.env['sale.order'].search([('analytic_account_id', 'in', analytic_accounts.ids), ('state', '=', 'sale')], order='create_date DESC')
+            for sale_order in sale_orders:
+                mapping[sale_order.analytic_account_id.id] = sale_order
```
Comment (dbeguin) [inline addons/sale/models/account_move.py]: use dict comprehension as you dict is empty

File: addons/sale/models/account_move.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            analytic_accounts_without_open_order = analytic_accounts.filtered(lambda account: not mapping.get(account.id))
+            if analytic_accounts_without_open_order:  # then, fill the blank with not open sales orders
+                sale_orders = self.env['sale.order'].search([('analytic_account_id', 'in', analytic_accounts_without_open_order.ids)], order='create_date DESC')
+            for sale_order in sale_orders:
+                mapping[sale_order.analytic_account_id.id] = sale_order
+
+        # map of AAL index with the SO on which it needs to be reinvoiced. Maybe be None if no SO found
+        result = {}
+        for move_line in self:
+            result[move_line.id] = mapping.get(move_line.analytic_account_id.id)
```
Comment (dbeguin) [inline addons/sale/models/account_move.py]: same : dict comprehension

File: addons/sale_expense/models/account_move.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+        """ For move lines created from expense, we override the normal behavior.
+            Note: if no SO but an AA is given on the expense, we will determine anyway the SO from the AA, using the same
+            mecanism as in Vendor Bills.
+        """
+        mapping_from_invoice = super(AccountMoveLine, self)._sale_determine_order()
+
+        mapping_from_expense = {}
+        for move_line in self.filtered(lambda move_line: move_line.expense_id):
+            if move_line.expense_id.sale_order_id:
+                mapping_from_expense[move_line.id] = move_line.expense_id.sale_order_id or None
```
Comment (dbeguin) [inline addons/sale_expense/models/account_move.py]: well here you could also use dict comprehension but for readability purpose.. you win

File: addons/sale/models/account_move.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+        amount = (self.credit or 0.0) - (self.debit or 0.0)
+
+        if self.product_id.expense_policy == 'sales_price':
+            return self.product_id.with_context(
+                partner=order.partner_id.id,
+                date_order=order.date_order,
+                pricelist=order.pricelist_id.id,
+                uom=self.product_uom_id.id
+            ).price
+        if unit_amount == 0.0:
```
Comment (tde-banana-odoo) [inline addons/sale/models/account_move.py]: Float compare ? (or float is zero that I think exists)

Comment (jem-odoo) [reply addons/sale/models/account_move.py]: Probably, but with which precision ? 
As the product uom is not required on account.move.line, should we use the rounding of the UoM and fallback on `dp.get_precision('Product Unit of Measure')`?

**Comment by jem-odoo** [comment]
@robodoo rebase-ff r+

**Pull request closed without merging**